include/gridtools/common/int_vector.hpp

```diff
@@ -0,0 +1,134 @@
+        struct value_t_merger;
+
+        template <template <class...> class Item, class Key, class... Vs>
+        struct value_t_merger<Item<Key, Vs>...> {
+            using type = Item<Key, std::decay_t<decltype((Vs{} + ...))>>;
+        };
+
+        template <class... T>
+        using merger_t = typename value_t_merger<T...>::type;
+
+        template <class Key>
+        struct add_f {
+            template <class... Ts>
+            GT_FUNCTION GT_CONSTEXPR decltype(auto) operator()(Ts const &...args) const {
+                return (host_device::at_key_with_default<Key, integral_constant<int, 0>>(args) + ...);
+            }
+        };
+
+        template <class First, class Second>
+        GT_FUNCTION GT_CONSTEXPR auto plus(First &&first, Second &&second) {
+            using merged_meta_map_t = meta::mp_make<int_vector_impl_::merger_t,
+                meta::concat<hymap::to_meta_map<First>, hymap::to_meta_map<Second>>>;
+            using keys_t = meta::transform<meta::first, merged_meta_map_t>;
+            using generators = meta::transform<add_f, keys_t>;
+            return tuple_util::host_device::generate<generators, hymap::from_meta_map<merged_meta_map_t>>(
+                std::forward<First>(first), std::forward<Second>(second));
+        }
+
+        template <class Vec, class Scalar>
+        GT_FUNCTION GT_CONSTEXPR auto multiply(Vec &&vec, Scalar scalar) {
+            return tuple_util::host_device::transform([scalar](auto v) { return v * scalar; }, std::forward<Vec>(vec));
+        }
+
+        template <class I, class Enable = void>
+        struct is_integral_zero : std::false_type {};
+
+        template <class I>
+        struct is_integral_zero<I,
+            std::enable_if_t<is_integral_constant<meta::second<I>>::value && meta::second<I>::value == 0>>
+            : std::true_type {};
+
+        template <class Key>
+        struct at_key_f {
+            template <class T>
+            GT_FUNCTION GT_CONSTEXPR decltype(auto) operator()(T const &arg) const {
+                return host_device::at_key<Key>(arg);
+            }
+        };
+
+        template <class Vec>
+        GT_FUNCTION GT_CONSTEXPR auto normalize(Vec &&vec) {
+            using filtered_map_t = meta::filter<meta::not_<is_integral_zero>::apply, hymap::to_meta_map<Vec>>;
+            using keys_t = meta::transform<meta::first, filtered_map_t>;
+            using generators = meta::transform<at_key_f, keys_t>;
+            return tuple_util::host_device::generate<generators, hymap::from_meta_map<filtered_map_t>>(
+                std::forward<Vec>(vec));
+        }
+    } // namespace int_vector_impl_
+
+    template <class T>
+    struct is_int_vector : std::true_type {
+    }; // TODO proper type trait: is_hymap and (is_integral or is_gr_integral_constant)
+
+    template <class T>
+    inline constexpr bool is_int_vector_v = is_int_vector<T>::value;
+
+    namespace int_vector {
+        using int_vector_impl_::multiply;
```

Comment:
I think you sometimes use this style to hide some complex implementation in the impl namespace. What's your rationale when to do it?

Reply:
Good question. Let me cover this topic here. Probably we should revive gridtools C++ style guide in some form an place the rationales there (applicable mostly for a new code, old code might violate):
- The directory structure of the `include` should be in sync with c++ namespaces. 
  - for any namespace `n` should be either a directory `n` or a header `n.cpp`
  - for any directory `dir` (that is not `common`), all c++ entities that are defined in that directory should be within the `dir` namespace. 
  - for `dir/common` directory all c++ entities are defined within the `dir` namespace
  - if directory has public subdirectories (`internal`, `detail`, `private`, `impl` etc. are considered not public), it should not have headers directly. The exceptions are umbrella headers (like `gridtools/meta.hpp`) which contains only `#include` directives pointed to the headers within the subdirectory  of the same name. This is kind of auxiliary recommendation. It allows to keep explorer pane in IDEs neat and compact.   
  - if some directory do not follow those rules for the reason, it is better to have `README.md` in it. Like in `gridtools/stencil`.
- headers should place all private c++ entities into implementation namespace. This namespace should be unique (two headers should not share the same implementation namespace). The usual pattern is to use `name::impl_` if the header   `name.hpp` manages its own namespace otherwise `name_impl_`. There are two patterns to deal with implementation namespaces:
  - define everything within the `impl_` and expose the public entities  to the outer namespace with `using`;
  - define public entities it the outer namespace and use internal stuff explicitly (`impl_::impl_foo` within `foo`).
 In some cases the second pattern doesn't because your design relies on ADL. If it is not the case, the choice should be driven by readability.
 Examples:
   - You have ten public entities in your header, only one public entity uses a private thing that is one liner -- go for the second pattern.
   - You have ten private entities and a single public one that directly uses all private things -- go for the first pattern.